testing/citest/tests/openstack_smoke_test.py

```diff
@@ -0,0 +1,258 @@
+    # because it scopes the context of our activities.
+    # pylint: disable=invalid-name
+    self.TEST_APP = bindings['TEST_APP']
+    self.TEST_STACK = bindings['TEST_STACK']
+    self.TEST_SECURITY_GROUP = 'sec_grp_'+ bindings['TEST_APP']
+
+
+  def create_app(self):
+    """Creates OperationContract that creates a new Spinnaker Application."""
+    contract = jc.Contract()
+    return st.OperationContract(
+        self.agent.make_create_app_operation(
+            bindings=self.bindings, application=self.TEST_APP,
+            account_name=self.bindings['SPINNAKER_OS_ACCOUNT']),
+        contract=contract)
+
+  def delete_app(self):
+    """Creates OperationContract that deletes a new Spinnaker Application."""
+    contract = jc.Contract()
+    return st.OperationContract(
+        self.agent.make_delete_app_operation(
+            application=self.TEST_APP,
+            account_name=self.bindings['SPINNAKER_OS_ACCOUNT']),
+        contract=contract)
+
+  def create_a_security_group(self):
+    """Creates OsContract for createServerGroup.
+
+    To verify the operation, we just check that the spinnaker security group
+    for the given application was created.
+    """
+    rules = [{
+        "fromPort":80,
+        "toPort":80,
+        "prevcidr":"0.0.0.0/0",
+        "cidr":"0.0.0.0/0",
+        "ruleType":"TCP",
+        "remoteSecurityGroupId":"",
+        "icmpType":-1,
+        "icmpCode":-1
+        },
+             {
+                 "fromPort":10,
+                 "toPort":65530,
+                 "prevcidr":"",
+                 "cidr":"",
+                 "ruleType":"UDP",
+                 "remoteSecurityGroupId":"SELF",
+                 "icmpType":-1,
+                 "icmpCode":-1
+             },
+             {
+                 "fromPort":1,
+                 "toPort":65535,
+                 "prevcidr":"",
+                 "cidr":"",
+                 "ruleType":"ICMP",
+                 "remoteSecurityGroupId":"SELF",
+                 "icmpType":12,
+                 "icmpCode":9}]
+
+    job = [{
+        "provider": "openstack",
+        "region": self.bindings['OS_REGION_NAME'],
+        "stack": self.TEST_STACK,
+        "description": "Test - create security group for {app}".format(
+            app=self.TEST_APP),
+        "detail": "",
+        "account": self.bindings['SPINNAKER_OS_ACCOUNT'],
+        "rules": rules,
+        "name": self.TEST_SECURITY_GROUP,
+        "securityGroupName": self.TEST_SECURITY_GROUP,
+        "cloudProvider": "openstack",
+        "type": "upsertSecurityGroup",
+        "user": self.bindings['TEST_OS_USERNAME']
+    }]
+
+    builder = os.OsContractBuilder(self.os_observer)
+
+    (builder.new_clause_builder(
+        'Security Group Created', retryable_for_secs=30)
+     .show_resource('security group', self.TEST_SECURITY_GROUP)
+     .contains_match({'name': jp.STR_SUBSTR(self.TEST_SECURITY_GROUP),
+                      'rules': jp.STR_SUBSTR("protocol='tcp'")
```

Comment:
This wont do what you think because it is handled as an OR not an AND. I'm looking into fixing the API.

Instead you want
.contains_pred_list([
   jc.DICT_MATCH(),
   jc.DICT_MATCH(),
   jc.DICT_MATCH()
])

That will be an AND of all the constraints.